core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -502,7 +509,7 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
                         UnassignedInfo primaryFailedUnassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.PRIMARY_FAILED,
                             "primary failed while replica initializing", null, 0, unassignedInfo.getUnassignedTimeInNanos(),
                             unassignedInfo.getUnassignedTimeInMillis(), false, AllocationStatus.NO_ATTEMPT);
-                        failShard(logger, replicaShard, primaryFailedUnassignedInfo, indexMetaData);
+                        failShard(logger, replicaShard, primaryFailedUnassignedInfo, indexMetaData, routingChanges);
                     }
                 }
             }
```

Comment:
It's hidden by the fold below (so I can't comment on it), but does th' `if (failedShard.relocating()) { ... }` branch need to make any changes to the routing changes, or are those handled by the `remove(targetShard)` and `removeRelocationSource(targetShard)` methods?